File: src/containers/gifts.ts
import type { Gift, Student } from "@prisma/client";
import { db } from "../db/client";

export type DetailedGift = Gift & {
  adoredBy: Student[];
  lovedBy: Student[];
  likedBy: Student[];
};

export class GiftContainer {
  private gifts: DetailedGift[] = [];

  async bootstrap(): Promise<void> {
    await this.reload();
  }

  async reload(): Promise<void> {
    this.gifts = await db.gift.findMany({
      include: {
        adoredBy: true,
        lovedBy: true,
        likedBy: true,
      },
      orderBy: {
        name: "asc",
      },
    });
  }

  getGifts(): Gift[] {
    return this.gifts;
  }

  getGiftWithName(name: string): DetailedGift | null {
    const finalName = name.toLowerCase();
    return (
      this.gifts.find((gift) => gift.name.toLowerCase() === finalName) || null
    );
  }

  findManyByName(name: string): DetailedGift[] {
    return this.gifts.filter((gift) =>
      gift.name.toLowerCase().includes(name.toLowerCase()),
    );
  }
}

export const giftContainer = new GiftContainer();

File: src/containers/gifts.rs
use crate::db::models::DetailedGift;

#[derive(Debug, Default)]
pub struct GiftContainer {
    gifts: Vec<DetailedGift>,
}

impl GiftContainer {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn reload(&mut self, gifts: Vec<DetailedGift>) {
        self.gifts = gifts;
    }

    pub fn get_gifts(&self) -> &[DetailedGift] {
        &self.gifts
    }

    pub fn get_gift_with_name(&self, name: &str) -> Option<&DetailedGift> {
        self.gifts
            .iter()
            .find(|g| g.name.eq_ignore_ascii_case(name))
    }

    pub fn find_many_by_name(&self, name: &str) -> Vec<&DetailedGift> {
        let lower = name.to_lowercase();
        self.gifts
            .iter()
            .filter(|g| g.name.to_lowercase().contains(&lower))
            .collect()
    }
}
